s3_storage_provider.py

```diff
@@ -76,41 +77,47 @@ def __init__(self, hs, config):
         if "secret_access_key" in config:
             self.api_kwargs["aws_secret_access_key"] = config["secret_access_key"]
 
+        self._s3_client = None
+
         threadpool_size = config.get("threadpool_size", 40)
-        self._download_pool = ThreadPool(
-            name="s3-download-pool", maxthreads=threadpool_size
-        )
-        self._download_pool.start()
+        self._s3_pool = ThreadPool(name="s3-pool", maxthreads=threadpool_size)
+        self._s3_pool.start()
 
         # Manually stop the thread pool on shutdown. If we don't do this then
         # stopping Synapse takes an extra ~30s as Python waits for the threads
         # to exit.
         reactor.addSystemEventTrigger(
-            "during", "shutdown", self._download_pool.stop,
+            "during", "shutdown", self._s3_pool.stop,
         )
 
+    def _get_s3_client(self):
+        s3 = self._s3_client
+        if not s3:
+            b3_session = boto3.session.Session()
+            self._s3_client = s3 = b3_session.client("s3", **self.api_kwargs)
+        return s3
+
     def store_file(self, path, file_info):
         """See StorageProvider.store_file"""
 
         def _store_file():
-            session = boto3.session.Session()
-            session.resource("s3", **self.api_kwargs).Bucket(self.bucket).upload_file(
+            self._get_s3_client().upload_file(
```

Comment:
I'm struggling to find good docs on the API for boto, but according to https://boto3.amazonaws.com/v1/documentation/api/latest/guide/s3-uploading-files.html I think this should work?